### ISSUE TITLE
Cross-domain authentication

### ISSUE THREAD

**Issue by tituspijean** [tituspijean/flarum-ext-auth-ssowat#2]
[Issue ](https://github.com/YunoHost-Apps/flarum_ynh/issues/18) from @jellium.

> This issue follows a [discussion on the YunoHost forum](https://forum.yunohost.org/t/flarum-forum-package/1971/49?u=jellium).
> 
> I successfully installed the flarum SSOwat extension with the following command:
> 
> `sudo -u www-data composer require tituspijean/flarum-ext-auth-ssowat:*@dev` 
> 
> and activated it in the Flarum extension panel.
> 
> After that, 
> 1) I cannot login directly from my YunoHost SSO authentication: I need to login manually in Flarum with my YnH login/password.
> 2) The `Login with SSOwat` button in the Flarum login popup brings me to a 404 error page, trying to reach: `https://domain.tld/yunohost/sso/?r=some 40 characters long hash`
> 3) When logged in and browsing the forum, I get the following error:
> `Oops! Something went wrong. Please reload the page and try again.` associated with DEBUG verbose output untitled `200 OK` which contains the content I am trying to see, formatted with JSON notations (I think?)
> 
> Thanks!

**Comment by tituspijean** [comment]
SSOwat is actually on an other domain than Flarum's. There is no way to detect where SSOwat is, so to change the redirection url we need to implement extension settings.

The `200 OK` error seems may be related to #1.

**Comment by jellium** [comment]
Follow up of [this thread](https://forum.yunohost.org/t/flarum-forum-package/1971/56).

I didn't notice the available settings for the extension! My bad :smile: 

We are progressing, I am sure.

The Login with SSOwat link turns the `Log In` "popup window" to what looks like the `Sign Up` "popup window. The `Username` area is filled with the YunoHost username under which I am SSO authenticated. Below it there is the corresponding YunoHost email address written in an none editable area. No password area is present. The confirm button is called `Sign Up` and finally below it there is `Already have an account? Log In`.

**Comment by tituspijean** [comment]
Yeah that's normal. Do you get an error on Sin up ?

**Comment by jellium** [comment]
From the browser's dev tool, please see the network log (only the errors) detailed below:

- GET https://subdomain2.domain.tld/ynhpanel.js [HTTP/1.1 404 Not Found 20ms]

_Response headers_:
**Connection** keep-alive**
Content-Encoding** gzip
**Content-Type** text/html
**Date** Sun, 12 Feb 2017 15:02:58 GMT
**Server** nginx
**Set-Cookie** SSOwAuthRedirect=;; Path=/yunohost/sso/; Max-Age=Thu, Jan 01 1970 00:00:00 UTC;
**Transfer-Encoding** chunked
**Vary** Accept-Encoding
**X-SSO-WAT** You've just been SSOed

_Request headers_:
**Accept** */*
**Accept-Encoding** gzip, deflate, br
**Accept-Language** en-US,en;q=0.5
**Connection** keep-alive
**Cookie** flarum_session=j6sv3j7lauergfn7s6fnh61ot5; SSOwAuthUser=jelle; SSOwAuthHash=ddc48c6647ecfde456af6fa2f8b23a6d; SSOwAuthExpire=1487516556.088
**DNT** 1
**Host** subdomain2.domain.tld
**Referer** https://subdomain2.domain.tld/
**User-Agent** Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:51.0) Gecko/20100101 Firefox/51.0

- POST XHR https://subdomain2.domain.tld/register [HTTP/1.1 422 Unprocessable Entity 405ms]

_Response Headers_:
**Connection** keep-alive
**Content-Length** 292
**Content-Type** application/vnd.api+json
**Date** Sun, 12 Feb 2017 15:03:38 GMT
**Server** nginx
**Set-Cookie** SSOwAuthRedirect=;; Path=/yunohost/sso/; Max-Age=Thu, Jan 01 1970 00:00:00 UTC;
flarum_session=j6sv3j7lauergfn7s6fnh61ot5; Path=/; HttpOnly
**X-CSRF-Token** YwjZ8KAbhm7NpxpPsE5TvhOx9dXAkvl0JcNzjvC3
**X-SSO-WAT** You've just been SSOed

_Request headers_:
**Accept-Language** en-US,en;q=0.5
**Connection** keep-alive
**Content-Length** 96
**Content-Type** application/json; charset=utf-8
**Cookie** flarum_session=j6sv3j7lauergfn7s6fnh61ot5; SSOwAuthUser=YnH SSO username; SSOwAuthHash=ddc48c6647ecfde456af6fa2f8b23a6d; SSOwAuthExpire=1487516556.088
**DNT** 1
**Host** subdomain2.domain.tld
**Referer** https://subdomain2.domain.tld/
**User-Agent** Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:51.0) Gecko/20100101 Firefox/51.0
**X-CSRF-Token** YwjZ8KAbhm7NpxpPsE5TvhOx9dXAkvl0JcNzjvC3

_Response_:
**Raw Data**
`{"errors":[{"status":"422","code":"validation_error","detail":"The username has already been taken.","source":{"pointer":"\/data\/attributes\/username"}},{"status":"422","code":"validation_error","detail":"The email has already been taken.","source":{"pointer":"\/data\/attributes\/email"}}]`

_TypeError: error.response is null forum-21f5938f.js:18634:19_
**request/<** https://subdomain2.domain.tld/assets/forum-21f5938f.js:18634:19
**fire/<** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2541:22
**thennable** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2510:5
**fire** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2530:4
**Deferred/self.reject** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2458:5
**finish/<** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2484:6
**map** self-hosted
**finish** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2480:4
**fire/</<** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2557:7
**thennable** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2510:5
**fire/<** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2554:6
**thennable** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2510:5
**fire** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2530:4
**Deferred/self.reject** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2458:5
**m.request/options.onerror** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2779:5
**createXhr/xhr.onreadystatechange** https://subdomain2.domain.tld/assets/forum-21f5938f.js:2661:6

**Comment by tituspijean** [comment]
Ah ! It seems that once you create a user without the extension, you can't log in with SSOwat afterwards. I will see if it's a bug or a feature on Flarum's side.

**Comment by tituspijean** [comment]
Actually I think that's a feature. It would be easy to take over an admin's account by using an external service with same username and email.
Hence we should create the initial admin with the extension during post-installation of Flarum. (or at least delete the installation user and create the admin through SSOwat).

**Comment by tituspijean** [comment]
Scratch, almost, that. I am always redirected to the sign up modal, even is the user was already created through SSOwat. Damn, it worked with the LDAP version. :(

**Comment by jellium** [comment]
I created a test user for which the Login with SSOwat link works, as you suspected. There have been a red banner with this "Oops, something..." and I had to refresh the page, to get rid of it and the "Sign Up" popup I used to login with SSOwat with, in order to be actually logged in. So it's aaalmost good =)

For my understanding: does it have to be with a Login with SSOwat to propagate the SSOwat authentication to the app? What makes it different from, say packages Nextcloud, Roundcube, Rainloop, Kanboard and so on, for which the authentication is "transparent"? I think it is much better not to even give the possibility to log in with other credentials with those of the parent YunoHost session, if SSOwat is used.

Thanks for your inputs and close attention to this issue anyway, because I really think Flarum can be a true great choice of forum app.
